addons/binding/org.openhab.binding.spotify/src/main/java/org/openhab/binding/spotify/internal/handler/SpotifyBridgeHandler.java

```diff
@@ -0,0 +1,605 @@
+
+            props.put(PROPERTY_SPOTIFY_USER, user);
+            updateProperties(props);
+            return user;
+        }
+        return "";
+    }
+
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+        active = true;
+        configuration = getConfigAs(SpotifyBridgeConfiguration.class);
+        oAuthService = oAuthFactory.createOAuthClientService(thing.getUID().getAsString(), SPOTIFY_API_TOKEN_URL,
+                SPOTIFY_AUTHORIZE_URL, configuration.clientId, configuration.clientSecret, SPOTIFY_SCOPES, true);
+        oAuthService.addAccessTokenRefreshListener(SpotifyBridgeHandler.this);
+        spotifyApi = new SpotifyApi(oAuthService, scheduler, httpClient);
+        handleCommand = new SpotifyHandleCommands(spotifyApi);
+        playingContextCache = new ExpiringCache<>(configuration.refreshPeriod, spotifyApi::getPlayerInfo);
+        playlistCache = new ExpiringCache<>(configuration.refreshPeriod, spotifyApi::getPlaylists);
+        devicesCache = new ExpiringCache<>(configuration.refreshPeriod, spotifyApi::getDevices);
+
+        // Start with update status by calling Spotify. If no credentials available no polling should be started.
+        if (pollStatus()) {
+            startPolling();
+        }
+    }
+
+    @Override
+    public List<Device> listDevices() {
+        final List<Device> listDevices = devicesCache.getValue();
+
+        return listDevices == null ? Collections.emptyList() : listDevices;
+    }
+
+    /**
+     * Scheduled method to restart polling in case polling is not running.
+     */
+    private synchronized void scheduledPollingRestart() {
+        try {
+            final boolean pollingNotRunning = pollingFuture == null || pollingFuture.isCancelled();
+
+            playingContextCache.invalidateValue();
+            playlistCache.invalidateValue();
+            devicesCache.invalidateValue();
+
+            if (pollStatus() && pollingNotRunning) {
+                startPolling();
+            }
+        } catch (RuntimeException e) {
+            logger.debug("Restarting polling failed: ", e);
+        }
+    }
+
+    /**
+     * This method initiates a new thread for polling the available Spotify Connect devices and update the player
+     * information.
+     */
+    private synchronized void startPolling() {
+        cancelSchedulers();
+        if (active) {
+            playingContextCache.invalidateValue();
+            playlistCache.invalidateValue();
+            devicesCache.invalidateValue();
+            pollingFuture = scheduler.scheduleWithFixedDelay(this::pollStatus, 0, configuration.refreshPeriod,
+                    TimeUnit.SECONDS);
+        }
+    }
+
+    /**
+     * Calls the Spotify API and collects user data. Returns true if method completed without errors.
+     *
+     * @return true if method completed without errors.
+     */
+    private boolean pollStatus() {
+        synchronized (scheduler) {
```

Comment:
What is the intention here, you share the `scheduler` object with a lot of other things, AFAIK you should share on class-local things. What about creating a separate `Object` instance to synchronise over, now it looks kinda of incorrect because a lot of usages of `scheduler` or not synced.

Reply:
Did you address this one?

Reply:
Never mind my phone confused me

Reply:
Introduced a new field that is used instead of the `scheduler` object.